crates/wasmtime/src/func.rs

```diff
@@ -1698,28 +1697,118 @@ macro_rules! impl_into_func {
+                /// by Cranelift, since Cranelift is generating raw function
+                /// calls directly to this function.
+                unsafe extern "C" fn wasm_to_host_shim<F, $($args,)* R>(
+                    vmctx: *mut VMContext,
+                    caller_vmctx: *mut VMContext,
+                    $( $args: $args::Abi, )*
+                ) -> R::Abi
+                where
+                    F: Fn(Caller<'_>, $( $args ),*) -> R + 'static,
+                    $( $args: WasmTy, )*
+                    R: WasmRet,
+                {
+                    let state = (*vmctx).host_state();
+                    // Double-check ourselves in debug mode, but we control
+                    // the `Any` here so an unsafe downcast should also
+                    // work.
+                    debug_assert!(state.is::<F>());
+                    let func = &*(state as *const _ as *const F);
+
+                    let store = wasmtime_runtime::with_last_info(|last| {
+                        last.and_then(Any::downcast_ref::<Store>)
+                            .cloned()
+                            .expect("Engine function called without thread state")
+                    });
+
+                    let ret = {
+                        panic::catch_unwind(AssertUnwindSafe(|| {
+                            func(
+                                Caller { store: &store, caller_vmctx },
+                                $( $args::from_abi($args, &store), )*
+                            )
+                        }))
+                    };
+
+                    // Note that we need to be careful when dealing with traps
+                    // here. Traps are implemented with longjmp/setjmp meaning
+                    // that it's not unwinding and consequently no Rust
+                    // destructors are run. We need to be careful to ensure that
+                    // nothing on the stack needs a destructor when we exit
+                    // abnormally from this `match`, e.g. on `Err`, on
+                    // cross-store-issues, or if `Ok(Err)` is raised.
+                    match ret {
+                        Err(panic) => wasmtime_runtime::resume_panic(panic),
+                        Ok(ret) => {
+                            // Because the wrapped function is not `unsafe`, we
+                            // can't assume it returned a value that is
+                            // compatible with this store.
+                            if !ret.compatible_with_store(&store) {
+                                // Explicitly drop all locals with destructors prior to raising the trap
+                                drop(store);
+                                drop(ret);
+                                raise_cross_store_trap();
+                            }
+
+                            ret.into_abi_for_ret(&store)
+                        }
+                    }
+                }
```

Comment:
The problem is these two shims differ in how the store is accessed (one from capture and the other from TLS); related to that is the associated instance state also differs.

As such, I don't think much could be shared here since I wouldn't want the `resume_panic` call moved to any other frame than the shim function itself.

Reply:
We could unify these two actually by fetching the `Store` from the same location, the location in TLS post-dates the original storage into the function's environment here. I think then the two could be shared?

Reply:
If we want to do that rather than sourcing it from the instance data, I'll make the shim shared.

Reply:
Yeah I think we may as well change to that. It also has the added benefit of where the host state can remain 0-sized if `Self` is 0-sized (as is the case with all wasi functions). Additionally I think you could actually go ahead and remove `into_func` and simply have `into_host_func`, and then `Func::wrap` would have a small amount more glue to handle the add-the-instance-to-the-store case?

Reply:
That works.  Thoughts on renaming the trait then?